### DISCUSSION TITLE
Cannot update Content stream with doc.update_stream

### DISCUSSION THREAD

**Discussion by SirishaGorasa** [pymupdf/PyMuPDF#1614] [Looking for help]
_**Please provide all mandatory information!**_

## Describe the bug (mandatory)
Update content stream erases everything.

## To Reproduce (mandatory)

- Take content stream of tagged PDF with tagged link i.e. link as marked content sequence in content stream with an MCID(Marked content identifier)
- Create a text file with the decoded content stream.
- Read a similar PDF which is tagged except for the link i.e. link not present in stream.
- update the stream of this PDF with the stream of previous PDF, along with the changes required in internal structure(object definitions).
``` python
import fitz

#Open the text file containing the Content Stream of tagged PDF

file_name=open("text-file.txt",'r')

file_read=file_name.read()

file_name.close()

#Open the untagged pdf of the same tagged version to which the above text file corresponds to

doc=fitz.open("./google_link.pdf")

# taken care of mapping proper xref objects
obj="""<<

  /K [ <<

        /Obj <<

  /A <<

  /S /URI

  /Type /Action

  /URI ([https://www.google.com)]()

  >>

  /BS <<

    /S /S

    /Type /Border

    /W 0

  >>

  /Border [ 0 0 0 ]

  /Rect [ 53.9978 642.966 137.874 653.526 ]

  /StructParent 3

  /Subtype /Link

  /Type /Annot

>>

        /Type /OBJR

      >> 3 ]

  /Pg 3 0 R

  /S /Link

>>"""

xref=doc.get_new_xref()

doc.update_object(xref,obj)

# 12 is the contents xref of PDF file
doc.update_stream(12,file_read,compress=True)

doc.save("updated_google_link.pdf",clean=True)
```
Content stream is completely empty in new file.

## Expected behavior (optional)
Updated content stream in the new PDF.

## Screenshots (optional)
Completely blank PDF.

## Your configuration (mandatory)
 - OS: Windows, 64 bit
 - Python: 3.8.3
- Name: PyMuPDF
- Version: 1.19.5
- Summary: PyMuPDF is a Python binding for the document renderer and toolkit MuPDF
- Home-page: https://github.com/pymupdf/PyMuPDF
- Author: Jorj McKie
- Author-email: jorj.x.mckie@outlook.de
- License: GNU AFFERO GPL 3.0

## Additional context (optional)
Trying to find if we can modify internal structure and corresponding object stream to tag element (link in this context)

**Comment by JorjMcKie** [comment]
When you save a stream, you must make sure that the xref already is a stream object.
If you are not sure or if the xref is new, you must specify ``new=True`` in ``update_stream``.
In addition for a new xref: before you can use it as a PDF `dict` object, you must initialize it as such (which you did, I believe).

So if you make a new xref and then update it with the object definition of another (old) xref, then this does not make the new xref a stream object! Even if old xref was a stream object.

**Comment by SirishaGorasa** [comment]
we aren't creating a new xref, we are updating the existing xref containing the object stream with new stream.

**Comment by JorjMcKie** [comment]
Still try with the `new` please.
If it indeed does not work, please reopen, and I need a minimal reproducer with code snippet and example file.

**Comment by SirishaGorasa** [comment]
I have tried creating a new xref and updating it with the new stream. Then, I have modified the internal structure to make use of the new stream to look for contents in similar fashion as mentioned below.
```text
 /Contents 45 0 R
```
Then I have saved the document into a new file but I don't the expected content stream after extracting and decoding the content stream.

**Comment by JorjMcKie** [comment]
we put this in `Discussions`.

**Comment by JorjMcKie** [comment]
You are not telling the full story: how does this relate to your previous problem?
What is the problem now?
Did you update the stream with `new`?
Did the previous problem go away?
Etc., etc.
Please be more explicit!

**Reply by SirishaGorasa** [reply]
yes, the previous problem did go after using a new xref to add the stream.
So, with new xref it was mandatory to use `new` to update stream. I did use it.

But I don't see the new stream I wish to see when I decode the content stream of PDF.

**Reply by SirishaGorasa** [reply]
### Scenario
- I have created 2 pdfs with the same content.
- The 2 pdfs have 2 paragraphs and a link embedded in first paragraph.
- The difference between the two pdfs is the first pdf (Say tagged.pdf) has all contents tagged. The second pdf (say tagged_pdf_untagged_link.pdf) has the 2 paragraphs tagged but the link is untagged.
**-Now I want to replicate the logical structure and content stream of tagged link into the tagged pdf with untagged link and see whether the untagged link gets tagged and hence readable by NVDA Screen Reader which is not happening** 
### This is the below code using which I want to update the content stream for an untagged link in a tagged PDF
```import fitz
#Open the text file containing the Content Stream of tagged PDF 
file_name=open("tagged-contentstream.txt",'r')
file_read=file_name.read()
file_name.close()
#Open the Tagged PDF with untagged pdf object i.e., Link 
doc=fitz.open("./taggedpdf_with_untagged_link.pdf")
content_stream_xref=doc.get_new_xref()
doc.update_object(content_stream_xref,"<<>>")
doc.xref_set_key(3,"/Contents","{} 0 R".format(content_stream_xref))
doc.update_stream(content_stream_xref,file_read,new=True)
obj_1="""
<<
  /A <<
  /S /URI
  /Type /Action
  /URI (https://www.google.com)
>>
  /BS <<
    /S /S
    /Type /Border
    /W 0
  >>
  /Border [ 0 0 0 ]
  /C [ 0 1 0 ]
  /P 3 0 R
  /Rect [ 313.787 677.765 392.264 686.884 ]
  /Subtype /Link
  /Type /Annot
>>
"""
obj1_xref=doc.get_new_xref()
doc.update_object(obj1_xref,obj_1)
obj_2="""
 <<
  /K [ <<
        /Obj {} 0 R
        /Type /OBJR
      >> 4 ]
  /Pg 3 0 R
  /S /Link
>>
""".format(obj1_xref)
obj2_xref=doc.get_new_xref()
doc.update_object(obj2_xref,obj_2)
doc.save("google-stream_new_updated1.pdf")

**Comment by JorjMcKie** [comment]
I have no experience with marked content. So I am of no help with this topic.
All I can say is that you can of course verify that the `/Contents` in question indeed has the right contents. Other than that ... sorry.

**Reply by SirishaGorasa** [reply]
Could you please add any references here or any contact which can help me in this case?

**Reply by JorjMcKie** [reply]
The PDF specification: https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf

**Reply by SirishaGorasa** [reply]
### Update Stream method not working
I am again facing the same issue that I have mentioned earlier. update_stream() method of PyMuPDF is not writing content stream into the new xref object I have created. (Refer to codebase).
``` python
#Open the text file containing the Content Stream of tagged PDF 
file_name=open("tagged-contentstream.txt",'r')
file_read=file_name.read()
file_name.close()
#Open the Tagged PDF with untagged pdf object i.e., Link 
doc=fitz.open("./taggedpdf_with_untagged_link.pdf")
content_stream_xref=doc.get_new_xref()
doc.update_object(content_stream_xref,"<<>>")
doc.xref_set_key(3,"Contents","{} 0 R".format(content_stream_xref))
doc.update_stream(content_stream_xref,file_read,new=True)
obj_1="""
<<
  /A <<
  /S /URI
  /Type /Action
  /URI (https://www.google.com)
>>
  /BS <<
    /S /S
    /Type /Border
    /W 0
  >>
  /Border [ 0 0 0 ]
  /C [ 0 1 0 ]
  /P 3 0 R
  /Rect [ 313.787 677.765 392.264 686.884 ]
  /Subtype /Link
  /Type /Annot
>>
"""
obj1_xref=doc.get_new_xref()
doc.update_object(obj1_xref,obj_1)
obj_2="""
 <<
  /K [ <<
        /Obj {} 0 R
        /Type /OBJR
      >> 4 ]
  /Pg 3 0 R
  /S /Link
>>
""".format(obj1_xref)
obj2_xref=doc.get_new_xref()
doc.update_object(obj2_xref,obj_2)
doc.save("google-stream_new_updated1.pdf")
```

**Reply by SirishaGorasa** [reply]
It was never solved. Its due to a typo in the update_object which made me think the issue was solved.

**Reply by JorjMcKie** [reply]
What type of object are you trying to give a `/Contents` key? If it is not a page, then it is **_illegal!_**
Links have no such thing for example.
Here is a file with a code snippet that creates a page and a stream object with nonsense content, that is then stored as `/Contents` in the page definition.
All this works just fine:
[demo.txt](https://github.com/pymupdf/PyMuPDF/files/8169793/demo.txt)

**Reply by SirishaGorasa** [reply]
Is there any way I can overwrite the content stream of my PDF with the content stream of another PDF? (Taking care of structural changes)

**Reply by JorjMcKie** [reply]
Here the log for updating an existing stream. Also works fine.
[demo2.txt](https://github.com/pymupdf/PyMuPDF/files/8169890/demo2.txt)

**Reply by JorjMcKie** [reply]
> (Taking care of structural changes)

Don't know what that may mean in detail. All the contents of the stream, every bit and byte, is your responsibility. If you do this right, every name is correct, every reference to things in the PDF ... then it will work right as you see.
What you seem to be trying to do definitely is hacky and error-prone ...
